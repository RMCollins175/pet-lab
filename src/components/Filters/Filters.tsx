import { FormEventHandler } from "react";

export interface FiltersProps {
  handleSearch: FormEventHandler<HTMLFormElement>;
  handleReset: FormEventHandler<HTMLFormElement>;
  loading: boolean;
}

export const Filters = ({
  handleSearch,
  loading,
  handleReset,
}: FiltersProps) => {
  return (
    <div className="min-w-1/5 ">
      <form
        onSubmit={handleSearch}
        onReset={handleReset}
        className="flex flex-col justify-between">
        <label
          className="flex justify-between items-center border border-gray-200 rounded p-2.5 text-lg"
          data-testid="tag-label">
          Tag:
          <input
            className="w-1/2 p-1 border border-gray-200 rounded bg-gray-100"
            type="text"
            name="tag"
          />
        </label>
        <label
          className="flex justify-between items-center border border-gray-200 rounded p-2.5 text-lg"
          data-testid="price-label">
          Price:
          <input
            className="w-1/2 p-1 border border-gray-200 rounded bg-gray-100"
            type="number"
            name="price"
            min="1"
            max="1000"
          />
        </label>
        <label
          className="flex justify-between items-center border border-gray-200 rounded p-2.5 text-lg"
          data-testid="subscription-label">
          Subscription:
          <select
            name="subscription"
            className="w-1/2 p-1 border border-gray-200 rounded bg-gray-100">
            <option></option>
            <option value="true">Yes</option>
            <option value="false">No</option>
          </select>
        </label>
        <button
          type="submit"
          className="bg-gray-200 border border-gray-200 rounded text-lg p-2.5 mt-5 cursor-pointer"
          disabled={loading}
          data-testid="search-button">
          Search
        </button>
        <input
          type="reset"
          className="bg-gray-200 border border-gray-200 rounded text-lg p-2.5 mt-5 cursor-pointer"
          value="Reset"
          data-testid="reset-button"></input>
      </form>
    </div>
  );
};
