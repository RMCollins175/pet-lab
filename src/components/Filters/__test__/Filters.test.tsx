import React from "react";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Filters } from "../Filters";

describe("Filters Component", () => {
  test("renders all form elements", () => {
    const mockHandleSearch = jest.fn();
    render(<Filters handleSearch={mockHandleSearch} loading={false} />);

    expect(screen.getByTestId("tag-label")).toBeInTheDocument();
    expect(screen.getByTestId("price-label")).toBeInTheDocument();
    expect(screen.getByTestId("subscription-label")).toBeInTheDocument();
    expect(screen.getByTestId("search-button")).toBeInTheDocument();
    expect(screen.getByTestId("reset-button")).toBeInTheDocument();
  });

  test("allows input to be entered", () => {
    const mockHandleSearch = jest.fn();
    render(<Filters handleSearch={mockHandleSearch} loading={false} />);

    const tagInput = screen.getByLabelText("Tag:") as HTMLInputElement;
    fireEvent.change(tagInput, { target: { value: "test tag" } });
    expect(tagInput.value).toBe("test tag");

    const priceInput = screen.getByLabelText("Price:") as HTMLInputElement;
    fireEvent.change(priceInput, { target: { value: 100 } });
    expect(priceInput.value).toBe("100");
  });

  test("submits form with correct data", async () => {
    const mockHandleSearch = jest.fn();
    render(<Filters handleSearch={mockHandleSearch} loading={false} />);

    fireEvent.change(screen.getByLabelText("Tag:"), {
      target: { value: "test tag" },
    });
    fireEvent.change(screen.getByLabelText("Price:"), {
      target: { value: 100 },
    });
    fireEvent.click(screen.getByTestId("search-button"));

    await waitFor(() => {
      expect(mockHandleSearch).toHaveBeenCalled();
    });
  });

  // test("disables search button when loading", () => {
  //   const mockHandleSearch = jest.fn();
  //   render(<Filters handleSearch={mockHandleSearch} loading={true} />);

  //   expect(screen.getByText("Search")).toBeDisabled();
  // });

  test("resets form fields", () => {
    const mockHandleSearch = jest.fn();
    render(<Filters handleSearch={mockHandleSearch} loading={false} />);

    const tagInput = screen.getByLabelText("Tag:") as HTMLInputElement;
    fireEvent.change(tagInput, { target: { value: "test tag" } });
    fireEvent.click(screen.getByTestId("reset-button"));

    expect(tagInput.value).toBe("");
  });
});

describe("Filters Component Snapshot", () => {
  it("should match the snapshot", () => {
    const mockHandleSearch = jest.fn();
    const { asFragment } = render(
      <Filters handleSearch={mockHandleSearch} loading={false} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
