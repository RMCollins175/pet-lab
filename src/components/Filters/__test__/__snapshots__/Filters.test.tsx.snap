// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Component Snapshot should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="min-w-1/5 pt-4"
  >
    <form
      class="flex flex-col justify-between"
    >
      <label
        class="flex justify-between items-center border border-gray-200 rounded p-2.5 text-lg"
        data-testid="tag-label"
      >
        Tag:
        <input
          class="w-1/2 p-1 border border-gray-200 rounded bg-gray-100"
          name="tag"
          type="text"
        />
      </label>
      <label
        class="flex justify-between items-center border border-gray-200 rounded p-2.5 text-lg"
        data-testid="price-label"
      >
        Price:
        <input
          class="w-1/2 p-1 border border-gray-200 rounded bg-gray-100"
          max="1000"
          min="1"
          name="price"
          type="number"
        />
      </label>
      <label
        class="flex justify-between items-center border border-gray-200 rounded p-2.5 text-lg"
        data-testid="subscription-label"
      >
        Subscription:
        <select
          class="w-1/2 p-1 border border-gray-200 rounded bg-gray-100"
          name="subscription"
        >
          <option />
          <option
            value="true"
          >
            Yes
          </option>
          <option
            value="false"
          >
            No
          </option>
        </select>
      </label>
      <button
        class="bg-gray-200 border border-gray-200 rounded text-lg p-2.5 mt-5 cursor-pointer"
        data-testid="search-button"
        type="submit"
      >
        Search
      </button>
      <input
        class="bg-gray-200 border border-gray-200 rounded text-lg p-2.5 mt-5 cursor-pointer"
        data-testid="reset-button"
        type="reset"
        value="Reset"
      />
    </form>
  </div>
</DocumentFragment>
`;
